src/wasm/syscalls_openmp.cpp

```diff
@@ -123,22 +124,24 @@ namespace wasm {
      */
     WAVM_DEFINE_INTRINSIC_FUNCTION(env, "__kmpc_fork_call", void, __kmpc_fork_call, I32 locPtr, I32 argc,
                                    I32 microtaskPtr, I32 argsPtr) {
-        // The source loc info seems useless, but can be accessed with:
-        // Runtime::Memory *memoryPtr = getExecutingModule()->defaultMemory;
-        // wasm_ident *wasmLoc = &Runtime::memoryRef<wasm_ident>(memoryPtr, locPtr);
-        // logger->debug("wasmLoc.psource = {}", wasmLoc->psource);
-
         const std::shared_ptr<spdlog::logger> &logger = util::getLogger();
         logger->debug("S - __kmpc_fork_call {} {} {} {}", locPtr, argc, microtaskPtr, argsPtr);
 
+        // The source location contains information on what sort of code is calling
+        // this fork call. These types are defined in runtime/src/kmp.h
+        Runtime::Memory *memoryPtr = getExecutingModule()->defaultMemory;
+        wasm_ident *wasmLoc = &Runtime::memoryRef<wasm_ident>(memoryPtr, locPtr);
+
+        if(wasmLoc->flags != KMP_IDENT_WORK_LOOP) {
+            logger->error("Unsupported loc flags: {}", wasmLoc->flags);
+            throw std::runtime_error("Unsupported calling type for fork");
+        }
+
         // Retrieve the microstask function from the table
         Runtime::Object *funcObj = Runtime::getTableElement(getExecutingModule()->defaultTable, microtaskPtr);
         Runtime::Function *func = Runtime::asFunction(funcObj);
         IR::FunctionType funcType = Runtime::getFunctionType(func);
 
-        // Get reference to module memory
-        Runtime::GCPointer<Runtime::Memory> &memoryPtr = getExecutingModule()->defaultMemory;
```

Comment:
Why drop the`GCPointer`?

Reply:
`GCPointer` helps WAVM to perform its own garbage collection. It's useful to track long-lived references to WAVM-related stuff, so that it'll warn us if we're not able to clean up the module once we're done with it. If we're just accessing a pointer in a transient manner (e.g. inside a short function like this), then there's no need to add the complexity.